delta_go/game_mechanics.py

```diff
@@ -0,0 +1,265 @@
+import pickle
+import random
+import sys
+from pathlib import Path
+from typing import Any, Callable, Dict, Optional, Tuple
+
+import numpy as np
+
+import pygame
+import torch
+from go_base import (
+    BLACK,
+    BOARD_SIZE,
+    PASS_MOVE,
+    WHITE,
+    all_legal_moves,
+    game_over,
+    int_to_coord,
+    is_move_legal,
+    play_move,
+    result,
+    score,
+)
+from gym.spaces import Box, Discrete
+from pygame import Surface
+from render import render_game
+from state import State
+from torch import nn
+
+HERE = Path(__file__).parent.resolve()
+
+# Hack as this won't pip install on replit
+sys.path.append(str(HERE / "PettingZoo"))
+
+
+ALL_POSSIBLE_MOVES = np.arange(BOARD_SIZE**2 + 1)
+
+# Visuals
+SCREEN_SIZE = (500, 500)
+
+# The komi to use is much debated. 7.5 seems to
+# generalise well for different board sizes
+# lifein19x19.com/viewtopic.php?f=15&t=17750
+# 7.5 is also the komi used in alpha-go vs Lee Sedol
+# (non-integer means there are no draws)
+
+KOMI = 7.5
+
+
+def transition_function(state: State, action: int) -> State:
+    return play_move(state, action, state.to_play)
+
+
+def reward_function(state: State) -> int:
+    if not game_over(state.recent_moves):
+        return 0
+    result_ = result(state.board, KOMI)
+    return result_ if state.player_color == BLACK else result_ * -1
+
+
+def is_terminal(state: State) -> bool:
+    return game_over(state.recent_moves)
+
+
+def play_go(
+    your_choose_move: Callable,
+    opponent_choose_move: Callable,
+    game_speed_multiplier: float = 1.0,
+    render: bool = True,
+    verbose: bool = False,
+) -> float:
+
+    env = GoEnv(
+        opponent_choose_move,
+        verbose=verbose,
+        render=render,
+        game_speed_multiplier=game_speed_multiplier,
+    )
+
+    state, reward, done, info = env.reset()
+    done = False
+    while not done:
+        action = your_choose_move(state=state)
+        state, reward, done, info = env.step(action)
+    return reward
+
+
+# TODO: Currently state.board is just relative to the colors (BLACK = 1, WHITE = -1)
+# Need to think about whether to change before giving to choose_moves
```

Comment:
Imo this is fine because they have all the info they need. If they want a blacks channel and a whites channel for their NN they can easily do that
```suggestion
```